src/computability/tm_computable.lean

```diff
@@ -0,0 +1,220 @@
+/-
+Copyright (c) 2020 Pim Spelier, Daan van Gent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pim Spelier, Daan van Gent.
+-/
+
+import computability.encoding
+import computability.turing_machine
+import data.polynomial.basic
+import data.polynomial.eval
+
+/-!
+# Computable functions
+
+This file contains the definition of a Turing machine with some finiteness conditions
+(bundling the definition of TM2 in turing_machine.lean), a definition of when a TM gives a certain
+output (in a certain time), and the definition of computability (in polytime or any time function)
+of a function between two types that have an encoding (as in encoding.lean).
+
+## Main theorems
+
+- `id_computable_in_poly_time` : a TM + a proof it computes the identity on a type in polytime.
+- `id_computable`              : a TM + a proof it computes the identity on a type.
+
+## Implementation notes
+
+To count the execution time of a Turing machine, we have decided to count the number of times the
+`step` function is used. Each step executes a statement (of type stmt); this is a function, and
+generally contains multiple "fundamental" steps (pushing, popping, so on). However, as functions
+only contain a finite number of executions and each one is executed at most once, this execution
+time is up to multiplication by a constant the amount of fundamental steps.
+-/
+
+open computability
+
+/-- A bundled Turing Machine of type 2 (as defined in turing_machine.lean), with an input and
+output stack, a main function, an initial state and some finiteness guarantees. -/
+structure fin_tm2 :=
+ {K : Type}
+ [K_decidable_eq : decidable_eq K] -- index type of stacks
+ (k₀ k₁ : K) -- input and output stack
+ (Γ : K → Type) -- type of stack elements
+ (Λ : Type) (main : Λ)  -- type of function labels
+ (σ : Type) (initial_state : σ) -- type of states of the machine
+ [σ_fin : fintype σ]
+ [Γk₀_fin : fintype (Γ k₀)]
+ (M : Λ → turing.TM2.stmt Γ Λ σ) -- the program itself, i.e. one function for every function label
+
+namespace fin_tm2
+/-- The type of statements (functions) corresponding to this TM. -/
+@[derive inhabited]
+def stmt (tm : fin_tm2 ) : Type := @turing.TM2.stmt tm.K tm.K_decidable_eq tm.Γ tm.Λ tm.σ
+
+/-- The type of configurations (functions) corresponding to this TM. -/
+def cfg (tm : fin_tm2 ) : Type := @turing.TM2.cfg tm.K tm.K_decidable_eq tm.Γ tm.Λ tm.σ
```

Comment:
I haven't looked closely, but the fact that you have to use `@` so many times here and below is maybe a sign that some implicit arguments are wrong?

Reply:
I don't think so? In a section in the namespace turing.TM2, there is a parameter `[decidable_eq K]` floating around; this parameter is implicit in all the definitions in this section. But when we want to use it, we can't seem to pass it explicitly; if we try, we get a `failed to synthesize type class instance for ⊢ decidable_eq tm.K` error. This is what ended up working; if there is a better way to do it, we'd be glad to hear about it! :)

(Note: the parameter tm.K could be replaced by an `_`, but we felt that wouldn't improve the readibility)

Reply:
You need to make an instance `instance (tm : fin_tm2) : decidable_eq tm.K := tm.K_decidable_eq`.

Reply:
Or `attribute [instance] tm.K_decidable_eq` after the definition.